cli/js/dispatch_minimal.ts

```diff
@@ -17,23 +21,57 @@ export interface RecordMinimal {
   opId: number; // Maybe better called dispatchId
   arg: number;
   result: number;
+  err?: {
+    kind: ErrorKind;
+    message: string;
+  };
+}
+
+export function readInt32LittleEndian(
+  buffer: Uint8Array,
+  offset: number
+): number {
+  return (
+    (buffer[offset + 3] << 24) |
+    (buffer[offset + 2] << 16) |
+    (buffer[offset + 1] << 8) |
+    buffer[offset]
+  );
 }
 
 export function recordFromBufMinimal(
   opId: number,
-  buf32: Int32Array
+  ui8: Uint8Array
 ): RecordMinimal {
-  if (buf32.length != 3) {
-    throw Error("Bad message");
+  const promiseId = readInt32LittleEndian(ui8, 0);
+  const arg = readInt32LittleEndian(ui8, 4);
+  const result = readInt32LittleEndian(ui8, 8);
+  let err;
+
+  if (arg < 0) {
+    const kind = result as ErrorKind;
+    const message = decoder.decode(ui8.slice(12));
+    err = { kind, message };
+  } else if (ui8.length != 12) {
+    err = { kind: ErrorKind.InvalidData, message: "Bad message" };
   }
+
   return {
-    promiseId: buf32[0],
+    promiseId,
     opId,
-    arg: buf32[1],
-    result: buf32[2]
+    arg,
+    result,
+    err
   };
 }
```

Comment:
This code is very much in the hot path for our most important benchmark. I'm all for landing this to see what happens, but I will revert if it causes a noticeable slow down.

Reply:
Yes I agree completely. I will try to optimize it further, but we can't leave out proper error handling (which is long overdue anyway)